---
title: "Alpha Scan"
output: html_document
date: '2023-07-06'
---
```{r lib}
require(tidyr)
require(ggplot2)
require(data.table)
require(dplyr)
require(stringr)
require(stringi)



getAlpha <- function(progu.df){
  avg.progu <- ungroup(progu.df)
  avg.progu <- avg.progu %>% 
    # group_by(ordering) %>% 
    # mutate(delta_delta=delta_pos - lag(delta_pos, default = delta_pos[1])) %>%
    # ungroup() %>%
    group_by(ngen) %>% 
    summarize(mean_delta=mean(delta_pos))

  avg.progu <- avg.progu %>%
    mutate(mean_delta_delta = mean_delta - lag(mean_delta, default = mean_delta[1], order_by = ngen))
  new_item <- avg.progu$mean_delta_delta
  N=length(new_item)
  # ALPHA
  x = 2:N
  model = lm(log(new_item[x])~log(x))
  alpha = abs(model$coefficients[2])
  return(alpha)
}


getAlphas <- function(progu){
  
  k <-min(progu$kval):(max(progu$kval))
  #for each kval, calculate the mean delta delta for each ngen and then get the alpha coefficient to describe that line
  alpha <- sapply(k,
                function(x){
                  print(x)
                  subdf <- filter(progu,kval==x)
                  avalue <- as.numeric(getAlpha(subdf))
                  print(avalue)
                  return(avalue)
                  })
return(data.frame(k,alpha))
}


```

```{bash, eval=FALSE}
codelib=/home/jbonnie1/dandd/dev-dandD/lib/
codehelp=/home/jbonnie1/dandd/dev-dandD/helpers
outdir=/scratch4/blangme2/jessica/dandd/iscience/HGSVC2
sketchdir=/scratch4/blangme2/jessica/dandd/scratch/HGSVC2/sketchdb
tag=femrel
cd ${outdir}
filelist=${outdir}/consensus_fastas_female.txt
# cp /data/blangme2/jessica/HGSVC2/consensus_fastas_female.txt ${outdir}/.
### This is command to produce 10 random haplotypes:
shuf -n 12 ${outdir}/consensus_fastas_female.txt > ${outdir}/female_random10.txt
# filelist=${outdir}/female_random10.txt

# Do k=2..32 with dashing 
${codelib}/dandd tree -s ${tag} -f ${filelist} -o ${outdir} -k 20 -c ${sketchdir} --ksweep --lowmem
dash_pick=${outdir}/femrel_40_dashing_dtree.pickle

${codelib}/dandd progressive -d ${dash_pick} -n 20 --ksweep --lowmem #-o ${outdir}/
progdash=${outdir}/${tag}_progu30_40_dashing_dtree.pickle

${codelib}/dandd info -d ${progdash} --ksweep

# Do k=2..50 with kmc
${codelib}/dandd tree -s ${tag} -f ${filelist} -o ${outdir} -k 20 -c ${sketchdir} --ksweep --mink 2 --maxk 21 --lowmem --exact
kmc_pick=${outdir}/femrel_40_kmc_dtree.pickle

${codelib}/dandd progressive -d ${kmc_pick} -n 30 --ksweep --mink 2 --maxk 21  --lowmem #-o ${outdir}/
progkmc=${outdir}/${tag}_progu30_40_kmc_dtree.pickle

rm ${sketchdir}/ngen*/k[2..18]/*.kmc_*
${codelib}/dandd tree -s ${tag} -f ${filelist} -o ${outdir} -k 20 -c ${sketchdir} --ksweep --mink 20 --maxk 50 --lowmem --exact
kmc_pick=${outdir}/femrel_40_kmc_dtree.pickle

${codelib}/dandd progressive -d ${kmc_pick} -n 30 --ksweep --mink 2 --maxk 50 --lowmem  --exact #-o ${outdir}/
progkmc=${outdir}/${tag}_progu30_40_kmc_dtree.pickle

${codelib}/dandd info -d ${progdash} --ksweep


```

## Mitochondria Graphing Experiment


```{bash mito}
codelib=/home/jbonnie1/dandd/dev-dandD/lib/
codehelp=/home/jbonnie1/dandd/dev-dandD/helpers
datadir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-fish_mito/dataset/assembled-fish_mito/

outdir=/scratch4/blangme2/jessica/dandd/iscience/fish_mito/
filelist=${outdir}/fish_list.txt


tag=fish-mito
${codelib}/dandd tree -s ${tag} -f ${filelist} -o ${outdir} -k 10 --exact
kmc_pick=/scratch4/blangme2/jessica/dandd/iscience/fish_mito/fish-mito_25_kmc_dtree.pickle
${codelib}/dandd progressive -d ${kmc_pick} -n 60 --ksweep --mink 2 --maxk 80 --lowmem
kprogu=/scratch4/blangme2/jessica/dandd/iscience/fish_mito/fish-mito_progu80_25_kmc_dtree.pickle
${codelib}/dandd kij -d ${kprogu} 


```

```{r mitor}

fishperm=80
fishfile<-paste0('/scratch4/blangme2/jessica/dandd/iscience/fish_mito/fish-mito_progu',fishperm,'_25_kmcsummary.csv')
mito <- fread(fishfile) %>% select(-command) %>% arrange(kval,ordering,ngen) %>% filter(kval > 6)
# mito <- fread("~/dandd/dev-dandD/example/assembled-fish_mito/fish_progu0_25_kmcsummary.csv") %>% arrange(kval,ordering,ngen)  %>% filter(kval > 6) 

mito.alpha <- bind_rows(lapply(seq(20,fishperm,20), function(x){
  tmp <- mito %>%
    filter(ordering <= x)
  tmp.alpha <- getAlphas(tmp) %>%
    mutate(nperm=x)
  return(tmp.alpha)
}))

arrange(mito.alpha,as.numeric(k)) %>%
ggplot(
  aes(y=alpha, x=as.numeric(k), color=as.factor(nperm))) + 
  theme_bw() + #legend.position = c(.85, .8),
  # theme(legend.title="", legend.position = c(.85, .8) ) +
  # ggtitle('Distance: 1 - Jaccard' ) +
  geom_line() +
  geom_hline(yintercept = 1, color="red")
  # geom_point(size=2,alpha=.7)  
  # facet_wrap(~species, nrow = 2) + 
  # scale_shape_manual(values=c(15,17,19,8)) +
  # scale_color_manual(values = c(gg_color_hue(4)[1:3],"red")) +
  # theme_bw() 

```
## AF Project Ecoli

```{bash ecoli}
codelib=/home/jbonnie1/dandd/dev-dandD/lib/
codehelp=/home/jbonnie1/dandd/dev-dandD/helpers
datadir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-ecoli/dataset/assembled-ecoli
outdir=/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/
filelist=${outdir}/ecoli_list.txt

tag=ecoli
${codelib}/dandd tree -s ${tag} -f ${filelist} -o ${outdir} -k 14 --exact
kmc_pick=/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/ecoli_29_kmc_dtree.pickle
$codelib/../../dev-dandD_backup/lib/dandd progressive -d ${kmc_pick} -n 20 --verbose --ksweep --mink 2 --maxk 55 --lowmem

dash_pick=/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/ecoli_progu20_29_dashing_dtree.pickle
kmc_pick=/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/ecoli_29_kmc_dtree.pickle


refdir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-ecoli/reference
species=ecoli
#cd $outdir
reftree=${refdir}/ecoli_ref-tree.newick
refdata=${refdir}/dataset.json

python ${codelib}/dandd kij -d ${kmc_pick} -o $outdir/ecoli --mink 8 --maxk 32 --ksweep --afproject

mode=dashing
#--phylo-ref ${refdata} --phylo-tree ${reftree} 
python ${codelib}/../helpers/AFproject.py -d ${dash_pick} -o $outdir/ecoli --aftuples ${outdir}/ecoli_29_${mode}_AFtuples.pickle --phylo-tree $afloc/reference/ecoli_ref-tree.newick --phylo-ref $afloc/reference/dataset.json --mink 2 --maxk 31

cat ${outdir}/stats/*stats > ${upperout}/${species}_kij.tsv


```

```{r ecor}

ecoliperm=40
ecolifile<-paste0('/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/ecoli_progu20_29_kmcsummary.csv_2_55')
ecoli <- fread(ecolifile) %>% select(-command) %>% arrange(kval,ordering,ngen) %>% filter(kval > 9 , kval < 60)#%>% filter(delta_pos>0) #%>% filter(kval > 53, kval < 70)
# mito <- fread("~/dandd/dev-dandD/example/assembled-fish_mito/fish_progu0_25_kmcsummary.csv") %>% arrange(kval,ordering,ngen)  %>% filter(kval > 6) 

ecoli.alpha <- bind_rows(lapply(seq(5,20,5), function(x){
  tmp <- ecoli %>%
    filter(ordering <= x)
  tmp.alpha <- getAlphas(tmp) %>%
    mutate(nperm=x)
  return(tmp.alpha)
}))

arrange(ecoli.alpha,as.numeric(k)) %>%
  filter(nperm == 15) %>%
ggplot(
  aes(y=alpha, x=as.numeric(k), color=as.factor(nperm))) + 
  theme_bw() + #legend.position = c(.85, .8),
  # theme(legend.title="", legend.position = c(.85, .8) ) +
  # ggtitle('Distance: 1 - Jaccard' ) +
  geom_line() +
  geom_hline(yintercept = 1, color="red") +
  geom_vline(xintercept = 14, color="blue")
  # geom_point(size=2,alpha=.7)  
  # facet_wrap(~species, nrow = 2) + 
  # scale_shape_manual(values=c(15,17,19,8)) +
  # scale_color_manual(values = c(gg_color_hue(4)[1:3],"red")) +
  # theme_bw() 

```
K=14 says open. k=12 says closed. Open more reasonable

## HGSVC2 Females with Related



```{r pressure, echo=FALSE}
maxperm=4

curfile<-paste0('/scratch4/blangme2/jessica/dandd/iscience/HGSVC2/femrel_progu',maxperm,'_40_dashingsummary.csv')
femrel <- fread(curfile) %>% select(-command) %>% arrange(kval,ordering,ngen) %>% filter(kval > 11)

femrel.alpha <- bind_rows(lapply(seq(20,maxperm,20), function(x){
  tmp <- femrel %>%
    filter(ordering <= x)
  tmp.alpha <- getAlphas(tmp) %>%
    mutate(nperm=x)
  return(tmp.alpha)
}))



arrange(femrel.alpha,as.numeric(k)) %>%
ggplot(
  aes(y=alpha, x=as.numeric(k), color=as.factor(nperm))) + 
  theme_bw() + #legend.position = c(.85, .8),
  # theme(legend.title="", legend.position = c(.85, .8) ) +
  # ggtitle('Distance: 1 - Jaccard' ) +
  geom_line() +
  geom_hline(yintercept = 1, color="red")



```

```{r pressure, echo=FALSE}
maxperm=10
curfile<-paste0('/scratch4/blangme2/jessica/dandd/iscience/HGSVC2/femrel_progu4_40_kmcsummary.csv')
femrel <- fread(curfile) %>% select(-command) %>% arrange(kval,ordering,ngen) %>% filter(kval > 32)

femrel.alpha <- bind_rows(lapply(seq(1,maxperm,1), function(x){
  tmp <- femrel %>%
    filter(ordering <= x)
  tmp.alpha <- getAlphas(tmp) %>%
    mutate(nperm=x)
  return(tmp.alpha)
}))



arrange(femrel.alpha,as.numeric(k)) %>%
ggplot(
  aes(y=alpha, x=as.numeric(k), color=as.factor(nperm))) + 
  theme_bw() + #legend.position = c(.85, .8),
  # theme(legend.title="", legend.position = c(.85, .8) ) +
  # ggtitle('Distance: 1 - Jaccard' ) +
  geom_line() +
  geom_hline(yintercept = 1, color="red")



```



```{bash treemake, eval=FALSE}
codelib=/home/jbonnie1/dandd/dev-dandD/lib
upperout=/scratch4/blangme2/jessica/dandd/RECOMB/AFproject

#ECOLI
datadir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-ecoli/dataset/assembled-ecoli
refdir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-ecoli/reference
species=ecoli
outdir=${upperout}/${species}
cd $outdir
reftree=${refdir}/ecoli_ref-tree.newick
refdata=${refdir}/dataset.json

#dashing
python ${codelib}/dandd_cmd.py tree -s ${species} --datadir ${datadir} --out ${outdir}
tree_pick=${outdir}/ecoli_29_dashing_dtree.pickle
python ${codelib}/dandd_cmd.py kij -d ${tree_pick} -o ${outdir}/${species} --mink 2 --maxk 32 --phylo-ref ${refdata} --phylo-tree ${reftree}
cat ${outdir}/stats/*stats > ${upperout}/${species}_kij.tsv

#kmc
species=ecoli_kmc
outdir=${upperout}/${species}
cd $outdir
python ${codelib}/dandd_cmd.py tree -s ${species} --datadir ${datadir} --out ${outdir} --exact
tree_pick=${outdir}/${species}_29_kmc_dtree.pickle
python ${codelib}/dandd_cmd.py kij -d ${tree_pick} -o ${outdir}/${species} --mink 2 --maxk 60 --phylo-ref ${refdata} --phylo-tree ${reftree}

cat ${outdir}/stats/*stats > ${upperout}/${species}_kij.tsv

#FISH MITO
datadir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-fish_mito/dataset/assembled-fish_mito/
refdir=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-fish_mito/reference
species=fish-mito
outdir=${upperout}/${species}
mkdir ${outdir}
cd $outdir
reftree=${refdir}/tree.Fischer2013.newick
refdata=${refdir}/dataset.json

python ${codelib}/dandd_cmd.py tree -s ${species} --datadir ${datadir} --out ${outdir}
tree_pick=${outdir}/fish-mito_25_dashing_dtree.pickle
python ${codelib}/dandd_cmd.py kij -d ${tree_pick} -o ${outdir}/${species} --mink 2 --maxk 32 --phylo-ref ${refdata} --phylo-tree ${reftree}
cat ${outdir}/stats/*stats > ${upperout}/${species}_kij.tsv

##FISH KMC

species=fish-mito_kmc
outdir=${upperout}/${species}
mkdir ${outdir}
cd $outdir
reftree=${refdir}/tree.Fischer2013.newick
refdata=${refdir}/dataset.json

python ${codelib}/dandd_cmd.py tree -s ${species} --datadir ${datadir} --out ${outdir} --exact
tree_pick=${outdir}/fish-mito_kmc_25_kmc_dtree.pickle
python ${codelib}/dandd_cmd.py kij -d ${tree_pick} -o ${outdir}/${species} --mink 2 --maxk 60 --phylo-ref ${refdata} --phylo-tree ${reftree}
cat ${outdir}/stats/*stats > ${upperout}/${species}_kij.tsv

afloc=/data/blangme2/jessica/AF-reference_datasets190511/genome-std-assembled-ecoli/
 python AFproject.py -d ${dash_pick} -o $outdir --aftuples ${outdir}/ecoli_29_dashing_AFtuples.pickle --phylo-tree $afloc/reference/ecoli_ref-tree.newick --phylo-ref $afloc/reference/dataset.json --mink 2 --maxk 31

python AFproject.py -d ${dash_pick} -o $outdir --aftuples ${outdir}/ecoli_29_dashing_AFtuples.pickle --phylo-tree $afloc/reference/ecoli_ref-tree.newick --phylo-ref $afloc/reference/dataset.json --mink 2 --maxk 32
```


```{r outgraph}
e.dash <-fread("/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/ecoli_kij.tsv") 
colnames(e.dash) <- c('dist', 'k', 'k1', 'k2', 'k12', 'nRF', 'RF', 'maxRF', 'repl')
e.dash$tool <- "Dashing"
e.kmc <-fread("/scratch4/blangme2/jessica/dandd/RECOMB/AFproject/ecoli_kmc_kij.tsv")
colnames(e.kmc) <- c('dist', 'k', 'k1', 'k2', 'k12', 'nRF', 'RF', 'maxRF', 'repl')
e.kmc$tool <- "KMC"

ecoli <- e.dash %>% #rbind(e.dash, e.kmc) %>% 
  mutate(unik=ifelse(k==0,k12,k), method=ifelse(k==0,'KIJ','Fixed-k J'), species="ecoli")

f.dash <-fread("/scratch4/blangme2/jessica/dandd/RECOMB/AFproject/fish-mito_kij.tsv")
colnames(f.dash) <- c('dist', 'k', 'k1', 'k2', 'k12', 'nRF', 'RF', 'maxRF', 'repl')
f.dash$tool <- "Dashing"

f.kmc <- fread("/scratch4/blangme2/jessica/dandd/RECOMB/AFproject/fish-mito_kmc_kij.tsv")
colnames(f.kmc) <- c('dist', 'k', 'k1', 'k2', 'k12', 'nRF', 'RF', 'maxRF', 'repl')
f.kmc$tool <- "KMC"

fish <- rbind(f.dash, f.kmc) %>% 
  mutate(unik=as.numeric(ifelse(k==0,k12,k)), method=ifelse(k==0,'KIJ','Fixed-k J'), species="fish")

afdf <- rbind(fish, ecoli)
afdf$tool_and_method <- paste(afdf$tool, afdf$method)
write.csv(afdf, file="../figures/AFproject/kij_j_df.csv", row.names = FALSE)


gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}


ggplot(
  ecoli, aes(x=as.numeric(unik), y=nRF)) + #color=tool_and_method, shape=tool_and_method)) + 
  theme_bw() + #legend.position = c(.85, .8),
  theme(legend.title="", legend.position = c(.85, .8) ) +
  ggtitle('Distance: 1 - Jaccard' ) +
  geom_point(size=2,alpha=.7) + 
  facet_wrap(~species, nrow = 2) + 
  scale_shape_manual(values=c(15,17,19,8)) +
  scale_color_manual(values = c(gg_color_hue(4)[1:3],"red")) +
  theme_bw() 



```


```{r graph_triangular}

scan("/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/phylip/ecoli_AF_29_dashing.sim.k23.phylip",what = "list",skip = 1) -> phy23
scan("/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/phylip/ecoli_AF_29_dashing.sim.k24.phylip",what = "list",skip = 1) -> phy24
scan("/scratch4/blangme2/jessica/dandd/iscience/ecoliAF/phylip/ecoli_AF_29_dashing.sim.kij.phylip",what = "list",skip = 1) -> phyl
matrix(nrow=26,ncol=26) -> output
output[upper.tri(output, diag = FALSE)] <- phyl
finally <- t(output)
finally[,1] -> rownames(finally)
colnames(finally) <- rownames(finally)
finally2 <- finally[-c(1),-c(1)]

finallong <- as.data.table(finally2,keep.rownames = TRUE) %>% pivot_longer(cols=-1,names_to=c("sample1"),values_to = "dist",values_drop_na = TRUE) %>% mutate(dist=as.numeric(dist)) 
 ggplot(finallong,mapping=aes(x=sample1,y=rn,fill=dist)) + geom_tile(color="white") + 
  theme_bw()+
  theme(axis.text.x=element_text(angle=45, vjust=1, size=10, hjust=1)) +
  coord_fixed() +
  labs(x="",y="") +
  theme_bw() +
   geom_raster(aes(x=sample1,y=rn,fill=dist)) +
   scale_fill_gradient('viridis')
```




